File: 25. Clone list with random pointers.cpp
/*
 Task: Clone a linked list with next and random pointers
 Time Complexity: O(n)
 Space Complexity: O(n) using a hash map
*/

#include <iostream>
#include <unordered_map>

using namespace std;

struct Node
{
    int data;
    Node* next;
    Node* random;

    Node(int d)
        : data(d)
        , next(nullptr)
        , random(nullptr)
    {
    }
};

Node* clone(Node* head)
{
    if (!head)
    {
        return nullptr;
    }

    unordered_map<Node*, Node*> m;
    Node* cur = head;
    while (cur)
    {
        m[cur] = new Node(cur->data);
        cur = cur->next;
    }

    cur = head;
    while (cur)
    {
        m[cur]->next = m[cur->next];
        m[cur]->random = m[cur->random];
        cur = cur->next;
    }

    return m[head];
}

int main()
{
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->random = head->next->next;
    head->next->random = head;

    Node* c = clone(head);
    cout << c->data << " " << c->next->data << " " << c->random->data << endl;
    return 0;
}

// sc tc

File: 25. Clone list with random pointers.py
"""
Task: Clone a linked list with next and random pointers
Time Complexity: O(n)
Space Complexity: O(n) using a hash map
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.random = None


def clone(head):
    if not head:
        return None

    old_to_new = {}
    cur = head
    while cur:
        old_to_new[cur] = Node(cur.data)
        cur = cur.next

    cur = head
    while cur:
        old_to_new[cur].next = old_to_new.get(cur.next)
        old_to_new[cur].random = old_to_new.get(cur.random)
        cur = cur.next

    return old_to_new[head]


if __name__ == '__main__':
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    head.random = head.next.next
    head.next.random = head
    c = clone(head)
    print(c.data, c.next.data, c.random.data)

# Task: Clone a linked list with next and random pointers
